runtime/include/atomics/standard/chpl-atomics.h

```diff
@@ -0,0 +1,295 @@
+} \
+static inline type atomic_fetch_sub_explicit_ ## type(atomic_ ## type * obj, type operand, memory_order order) { \
+  type old_val = *obj; \
+  type new_val; \
+  do { \
+    new_val = old_val - operand; \
+  } while (!atomic_compare_exchange_weak_explicit(obj, &old_val, new_val, order, order)); \
+  return old_val; \
+} \
+static inline type atomic_fetch_sub_ ## type(atomic_ ## type * obj, type operand) { \
+  return atomic_fetch_sub_explicit_ ## type(obj, operand, memory_order_seq_cst); \
+}
+
+
+// Actually declare the atomics for integer and real types using the above macros
+DECLARE_ATOMICS_BASE(bool, chpl_bool);
+DECLARE_ATOMICS_EXCHANGE_OPS(bool, chpl_bool);
+
+
+#define DECLARE_ATOMICS(type) \
+  DECLARE_ATOMICS_BASE(type, type) \
+  DECLARE_ATOMICS_EXCHANGE_OPS(type, type); \
+  DECLARE_ATOMICS_FETCH_OPS(type)
+
+DECLARE_ATOMICS(int_least8_t);
+DECLARE_ATOMICS(int_least16_t);
+DECLARE_ATOMICS(int_least32_t);
+DECLARE_ATOMICS(int_least64_t);
+DECLARE_ATOMICS(uint_least8_t);
+DECLARE_ATOMICS(uint_least16_t);
+DECLARE_ATOMICS(uint_least32_t);
+DECLARE_ATOMICS(uint_least64_t);
+
+// On netbsd the DECLARE_ATOMICS macro doesn't work for uintptr_t. From gbt:
+// The root of the problem is the fact the on netbsd <stdint.h> (indirectly via
+// <sys.stdint.h>) #defines uintptr_t as __uintptr_t.  (__uintptr_t is in turn
+// typedef'd as unsigned long, but that doesn't matter to us.)  The C standard
+// (6.3.10.1(1) for C99) says that the actual arguments in a macro invocation
+// are themselves macro-expanded before being substituted into the replacement
+// text, unless they are preceded by a # or ## token.  In our case, the "type"
+// formal argument of DECLARE_ATOMICS_BASE has a ## before it in the
+// replacement text, so uintptr_t is not macro-expanded and the concatenation
+// produces the expected atomic_uintptr_t typedef name.  But in the case of
+// DECLARE_ATOMICS there is neither a # or ## before "type" in the replacement
+// text, so the uintptr_t actual becomes __uintptr_t via macro expansion before
+// DECLARE_ATOMICS_BASE is invoked, and we get a syntax error on the resulting
+// atomic___uintptr_t because it's not a typedef type.
+DECLARE_ATOMICS_BASE(uintptr_t, uintptr_t);
+DECLARE_ATOMICS_EXCHANGE_OPS(uintptr_t, uintptr_t);
+DECLARE_ATOMICS_FETCH_OPS(uintptr_t);
+
+
+#define DECLARE_REAL_ATOMICS(type) \
+  DECLARE_REAL_ATOMICS_BASE(type) \
+  DECLARE_REAL_ATOMICS_EXCHANGE_OPS(type) \
+  DECLARE_REAL_ATOMICS_FETCH_OPS(type)
+
+DECLARE_REAL_ATOMICS(_real32);
+DECLARE_REAL_ATOMICS(_real64);
+
+
+#undef DECLARE_ATOMICS_BASE
+#undef DECLARE_ATOMICS_EXCHANGE_OPS
+#undef DECLARE_ATOMICS_FETCH_OPS
+#undef DECLARE_REAL_ATOMICS_BASE
+#undef DECLARE_REAL_ATOMICS_EXCHANGE_OPS
+#undef DECLARE_REAL_ATOMICS_FETCH_OPS
+#undef DECLARE_ATOMICS
+#undef DECLARE_REAL_ATOMICS
+
+#endif // _chpl_atomics_h_
+
+/*
+ * Some misc notes:
+ *
+ *  - Our interface for the atomic_compare_exchange_* functions is slightly
```

Comment:
This is mentioned in the top level atomics README too right? If so, we can probably drop it from being in here as well, particularly since I think we included it in the atomics wish list jira issue